src/common/providers/identity.ts

```diff
@@ -34,21 +58,263 @@ export type UserRecord = firebase.auth.UserRecord;
+  claims?: { [key: string]: any };
+  idToken?: string;
+  accessToken?: string;
+  refreshToken?: string;
+  expirationTime?: string;
+  secret?: string;
+  providerId: string;
+  signInMethod: string;
+}
+
+/** Defines the auth event context for blocking events */
+export interface AuthEventContext extends EventContext {
+  locale?: string;
+  ipAddress: string;
+  userAgent: string;
+  additionalUserInfo?: AdditionalUserInfo;
+  credential?: Credential;
+}
+
+/** The handler response type for beforeCreate blocking events */
+export interface BeforeCreateResponse {
+  displayName?: string;
+  disabled?: boolean;
+  emailVerified?: boolean;
+  photoURL?: string;
+  customClaims?: object;
+}
+
+/** The handler response type for beforeSignIn blocking events */
+export interface BeforeSignInResponse extends BeforeCreateResponse {
+  sessionClaims?: object;
+}
+
+/** @internal */
+interface DecodedJwtMetadata {
+  creation_time?: number;
+  last_sign_in_time?: number;
+}
+
+/** @internal */
+interface DecodedJwtUserInfo {
+  uid: string;
+  display_name?: string;
+  email?: string;
+  photo_url?: string;
+  phone_number?: string;
+  provider_id: string;
+}
+
+/** @internal */
+export interface DecodedJwtMfaInfo {
+  uid: string;
+  display_name?: string;
+  phone_number?: string;
+  enrollment_time?: string;
+  factor_id?: string;
+}
+
+/** @internal */
+interface DecodedJwtEnrolledFactors {
+  enrolled_factors?: DecodedJwtMfaInfo[];
+}
+
+/** @internal */
+export interface DecodedJwtUserRecord {
+  uid: string;
+  email?: string;
+  email_verified?: boolean;
+  phone_number?: string;
+  display_name?: string;
+  photo_url?: string;
+  disabled?: boolean;
+  metadata?: DecodedJwtMetadata;
+  password_hash?: string;
+  password_salt?: string;
+  provider_data?: DecodedJwtUserInfo[];
+  multi_factor?: DecodedJwtEnrolledFactors;
+  custom_claims?: any;
+  tokens_valid_after_time?: number;
+  tenant_id?: string;
+  [key: string]: any;
+}
+
+/** @internal */
+export interface DecodedJwt {
+  aud: string;
+  exp: number;
+  iat: number;
+  iss: string;
+  sub: string;
+  event_id: string;
+  event_type: string;
+  ip_address: string;
+  user_agent?: string;
+  locale?: string;
+  sign_in_method?: string;
+  user_record?: DecodedJwtUserRecord;
+  tenant_id?: string;
+  raw_user_info?: string;
+  sign_in_attributes?: {
+    [key: string]: any;
+  };
+  oauth_id_token?: string;
+  oauth_access_token?: string;
+  oauth_refresh_token?: string;
+  oauth_token_secret?: string;
+  oauth_expires_in?: number;
+  [key: string]: any;
+}
+
 /**
  * Helper class to create the user metadata in a UserRecord object
  */
-export class UserRecordMetadata implements firebase.auth.UserMetadata {
+export class UserRecordMetadata implements UserMetadata {
   constructor(public creationTime: string, public lastSignInTime: string) {}
 
   /** Returns a plain JavaScript object with the properties of UserRecordMetadata. */
-  toJSON() {
+  toJSON(): object {
```

Comment:
I'm surprised you didn't get a warning for using a type as loose as `object` which can even include `undefined` IIRC. I'd probably make this return `Recrod<string, string>` or `{ creationTime: string, lastSignInTime: string }`

Edit: can't you say this returns `UserMetadata`?

Reply:
Changed the return type to `AuthUserMetadata`